create-derivative-topic/src/create_derivative_topic/utils.py

```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import pandas as pd
+
+def get_tracking_error(df_pos, df_sp):
+    """
+    Computes time-synchronized tracking error between vehicle position and setpoints.
+    
+    This function aligns the lower-frequency setpoint data to the higher-frequency 
+    position data using a forward-fill. It accounts for flight mode transitions 
+    where setpoints may be discontinuous.
+    
+    Returns a DataFrame with:
+    - err_x, err_y, err_z: Raw error per axis (NED frame).
+    - horiz_error: 2D Euclidean distance (sqrt(dx^2 + dy^2)).
+    - vert_error: Absolute vertical distance (abs(dz)).
+    """
+    # 1. Prepare Timestamps
+    for df in [df_pos, df_sp]:
+        # Ensure the index is datetime
+        if not isinstance(df.index, pd.DatetimeIndex):
+            df.index = pd.to_datetime(df.index)
+        
+        # Localize to UTC
+        if df.index.tz is None:
+            df.index = df.index.tz_localize('UTC')
+            
+        df.sort_index(inplace=True)
+    
+    # 2. Resample & Align
+    # We use df_pos as the "Master Clock" because it usually has a higher frequency.
+    # .reindex(method='ffill') carries the last known setpoint forward to match 
+    # the current position timestamp.
+    df_sp_aligned = df_sp.reindex(df_pos.index, method='ffill')
```

Comment:
If helpful, here's my take on signal alignment: https://github.com/roboto-ai/roboto-lerobot-actions/blob/main/mcap-to-lerobot/src/mcap_to_lerobot/lerobot.py#L49-L66